OutOfSchool/OutOfSchool.WebApi/Services/ProviderAdminService.cs

```
@@ -61,7 +65,7 @@ public class ProviderAdminService : CommunicationService, IProviderAdminService
         if (!hasAccess)
         {
             Logger.LogError("User(id): {UserId} doesn't have permission to create provider admin", userId);
-
+            apiErrorService.AddApiError(ApiErrorsTypes.ProviderAdmin.Creation.UserDontHavePermission(userId));
             return new ErrorResponse
```

Is it possible to extend ErrorResponse? Instead of `HttpStatusCode` and `Message`, it would return `HttpStatusCode` and `ApiError` or even better `List<ApiError>`. This way it directly contains all necessary information.

Based on the code base, we have both Either flow and Exceptions flow, so I'd like to offer some idea for the second one.
We could create some custom exception (let's say called "OperationException") that contain the same `HttpStatusCode` and `List<ApiError>`. This would allow us to catch it on the middleware side and return appropriate response to user. If needed, you could catch it earlier on controller side (don't know why).
This suggestion based on the already existing services throwing `InvalidOperationException`, `ArgumentOutOfRange`, ... with messages inside caught by controller level.

@DmyMi

---

@strivitech I'm not Dmytro, but have a question. 

Why don't use exceptions control flow at all? As we already have Either monad, developers get used to use it, so let's avoid exceptions wherever possible. 

Ofc exceptions are simple enough for call stack unrolling (and skipping few stack frames, till the first catch), but at same time they give lack of clarity, and provoke to grow your own "schizophrenia" for error handling. Also developers are used to use all "try catch bad practices" - to catch all exceptions, use `try`/`catch` when simple `if` is possible, therefore, maybe, it's time to forget them )

Summing up, let's do error handling in only way not both ways. What do you think?

---

@VadymLevkovskyi I think, ideally, it's good practice to have it one way. I can't say that using Either monads always better as it adds more "checks" on all nodes of chain calls and sometimes I would be glad to throw 😄. I agree that Either approach gives more clarity as it directly describes how logic works.

So, basically, I offered some workaround how to adapt existing code to use new ApiError classes.  I guess it's simpler to start with than change everything on Either. Can be wrong though.